src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -1,47 +1,173 @@
+
+        private T value;
+        private Node<T> prev;
+        private Node<T> next;
+
+        Node(T data) {
+            this.value = data;
+            this.prev = null;
+            this.next = null;
+        }
+    }
+
     @Override
     public void add(T value) {
+        Node<T> newNode = new Node<>(value);
+        if (head == null) {
+            head = newNode;
+            tail = newNode;
+
+        } else {
+            tail.next = newNode;
+            newNode.prev = tail;
+            tail = newNode;
+        }
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        if (index < 0 || index > size) {
+            throw new IndexOutOfBoundsException("Index out of bounds");
+        }
+
+        Node<T> newNode = new Node<>(value);
+
+        if (index == 0) {
+            newNode.next = head;
+            head = newNode;
+        } else {
+            Node<T> current = head;
+            for (int i = 0; i < index - 1; i++) {
+                current = current.next;
+            }
+            newNode.next = current.next;
+            current.next = newNode;
+        }
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T item : list) {
+            add(item);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index out of bounds");
+        }
+
+        Node<T> current = head;
+        for (int i = 0; i < index; i++) {
+            current = current.next;
+        }
+        return current.value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index out of bounds");
+        }
+
+        Node<T> current = head;
+        for (int i = 0; i < index; i++) {
+            current = current.next;
+        }
+
+        T oldValue = current.value;
+        current.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        if (index < 0 || index >= size) {
+            throw new IndexOutOfBoundsException("Index out of bounds");
+        }
+
+        Node<T> current = head;
+        Node<T> prev = null;
+
+        if (index == 0) {
+            T value = head.value;
+            head = head.next;
+            size--;
+            return value;
+        }
+
+        for (int i = 0; i < index; i++) {
+            prev = current;
+            current = current.next;
+        }
+
+        prev.next = current.next;
+        size--;
+        return current.value;
     }
 
     @Override
     public boolean remove(T object) {
+        if (head == null) {
+            return false;
+        }
+
+        Node<T> current = head;
+        while (current != null) {
+            if (Objects.equals(current.value, object)) {
```

Comment:
Consider using the `unlink(Node node)` method here to handle the removal logic. This will help you avoid repeating the unlinking logic.

Reply:
ok, done